### PR TITLE
Add a type-erased shared pointer

### PR DESCRIPTION
**Pull request by pyohannes**
I'm submitting this to make some headway on the `TracerProvider` PR: this adds a minimal type-erased `shared_ptr`, based on a [sketch by Ryan](https://gist.github.com/rnburn/886f677c79a43df64fa95b495a368524).

### CODE REVIEW

File: api/include/opentelemetry/nostd/shared_ptr.h [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+  {
+#if __EXCEPTIONS
+    try
+#endif
+    {
+      std::shared_ptr<T> ptr_(ptr);
+      new (buffer_.data) shared_ptr_wrapper{std::move(ptr_)};
+    }
+#if __EXCEPTIONS
+    catch (const std::bad_alloc &)
```
Comment (rnburn) [inline api/include/opentelemetry/nostd/shared_ptr.h]: Could we either remove the noexcept and let the constructor throw? or remove the try/catch and let it terminate on bad_alloc?

Comment (pyohannes) [reply api/include/opentelemetry/nostd/shared_ptr.h]: Done, I removed the try/catch and `noexcept`.

File: api/include/opentelemetry/nostd/shared_ptr.h [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+  element_type &operator*() const noexcept { return *wrapper().Get(); }
+
+  pointer operator->() const noexcept { return wrapper().Get(); }
+
+  operator bool() const noexcept { return wrapper().Get() != nullptr; }
+
+  pointer get() const noexcept { return wrapper().Get(); }
+
+  void swap(shared_ptr &other) noexcept { std::swap(buffer_, other.buffer_); }
```
Comment (rnburn) [inline api/include/opentelemetry/nostd/shared_ptr.h]: Bitwise swapping like this for a non-trivial type is guaranteed to work.

Comment (pyohannes) [reply api/include/opentelemetry/nostd/shared_ptr.h]: Probably you're right, who knows what kinds of `std::shared_ptr` will be wrapped. I replaced it with a three-way move.

File: api/include/opentelemetry/nostd/shared_ptr.h [hunk 3]
```diff
@@ -0,0 +1,181 @@
+#pragma once
+
+#include <memory>
+
+namespace opentelemetry
```
Comment (g-easy) [inline api/include/opentelemetry/nostd/shared_ptr.h]: `OPENTELEMETRY_BEGIN_NAMESPACE` ?
:)

File: api/test/nostd/sp [hunk 4]
```diff
@@ -0,0 +1,178 @@
+#include "opentelemetry/nostd/shared_ptr.h"
```
Comment (g-easy) [inline api/test/nostd/sp]: Delete this file?

File: api/include/opentelemetry/nostd/shared_ptr.h (hunk 3 above)
Comment (pyohannes) [reply api/include/opentelemetry/nostd/shared_ptr.h]: Done.

File: api/test/nostd/sp (hunk 4 above)
Comment (pyohannes) [reply api/test/nostd/sp]: Sorry for that - it's deleted now.

**Comment by pyohannes** [comment]
I think this should be ready to merge.

**Pull request merged**